aten/src/ATen/native/cuda/Sorting.cu

```diff
@@ -0,0 +1,431 @@
+      &median);
+
+  valuesSliceStart[0] = median;
+
+  // Find the median index
+  for (index_t i = threadIdx.x; i < inputSliceSize; i += blockDim.x) {
+    scalar_t val = doLdg(&inputSliceStart[i * inputWithinSliceStride]);
+    if (val == median ||
+        (THCNumerics<scalar_t>::isnan(val) &&
+         THCNumerics<scalar_t>::isnan(median))) {
+      indicesSliceStart[0] = i;
+      break;
+    }
+  }
+}
+
+struct KthValueLauncher {
+  int64_t k;
+
+  KthValueLauncher(int64_t k) : k(k) {}
+
+  template <typename scalar_t, typename index_t, int all_dims>
+  inline void launch(
+      cuda::detail::TensorInfo<scalar_t, index_t> values_info,
+      int collapse_values_dim,
+      cuda::detail::TensorInfo<int64_t, index_t> indices_info,
+      int collapse_indices_dim,
+      cuda::detail::TensorInfo<scalar_t, index_t> self_info,
+      int collapse_self_dim,
+      int64_t num_slices,
+      int64_t slice_size) {
+    dim3 grid;
+    if (!getGridFromTiles(num_slices, grid)) {
+      AT_ERROR("slices are too many");
+    }
+
+    dim3 block(std::min(
+        THCRoundUp(slice_size, (int64_t)C10_WARP_SIZE), (int64_t)1024));
+    auto stream = at::cuda::getCurrentCUDAStream();
+    gatherKthValue<scalar_t, index_t, all_dims><<<grid, block, 0, stream>>>(
+        self_info,
+        slice_size,
+        k,
+        num_slices,
+        /* The actual dimension that the k-selection is running in */
+        /* may have changed from collapseDims() */
+        self_info.strides[collapse_self_dim],
+        values_info,
+        indices_info);
+  }
+};
+
+struct MedianLauncher {
+  bool ignore_nan;
+
+  MedianLauncher(bool ignore_nan) : ignore_nan(ignore_nan) {}
+
+  template <typename scalar_t, typename index_t, int all_dims>
+  inline void launch(
+      cuda::detail::TensorInfo<scalar_t, index_t> values_info,
+      int collapse_values_dim,
+      cuda::detail::TensorInfo<int64_t, index_t> indices_info,
+      int collapse_indices_dim,
+      cuda::detail::TensorInfo<scalar_t, index_t> self_info,
+      int collapse_self_dim,
+      int64_t num_slices,
+      int64_t slice_size) {
+    dim3 grid;
+    if (!getGridFromTiles(num_slices, grid)) {
+      AT_ERROR("slices are too many");
+    }
+
+    dim3 block(std::min(
+        THCRoundUp(slice_size, (int64_t)C10_WARP_SIZE), (int64_t)1024));
+    auto stream = at::cuda::getCurrentCUDAStream();
+    gatherMedian<scalar_t, index_t, all_dims><<<grid, block, 0, stream>>>(
+        values_info,
+        indices_info,
+        self_info,
+        slice_size,
+        num_slices,
+        self_info.strides[collapse_self_dim],
+        ignore_nan);
+  }
+};
+
+template <typename scalar_t>
+void kthvalue_cuda_template(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t k,
+    int64_t dim_,
+    bool keepdim) {
+  int64_t dim = maybe_wrap_dim(dim_, self.dim());
+  int64_t slicesize = self.size(dim);
+  // FIXME: This seems bogus, I only do this because it was the old behaviour.
```

Comment:
You're correct. When a tensor has zero elements but the reduction is for a dimension whose size isn't zero then an empty tensor should be returned.

Reply:
This wasn't written by me, but as I mentioned above, we should fix a standard way of handling this for all operations with similar behavior like torch.min, torch.max etc.

Reply:
And then find some way to document or enforce it.

Reply:
Shouldn't we update this, too?

Reply:
I'll update this in another PR